send_email_messages/frontend/refresh-queue.js

```diff
@@ -0,0 +1,170 @@
+    messagesToCreate: [],
+    warnings: []
+  }
+  const allDeliveries = (await base.getTable("Deliveries").selectRecordsAsync()).records;
+  const allMessages = (await base.getTable("Messages").selectRecordsAsync()).records;
+  const allVolunteers = (await base.getTable("Volunteers").selectRecordsAsync()).records;
+
+  const EMAIL_TYPES = globalConfig.get('email_types');
+
+  for (const delivery of allDeliveries) {
+    const deliveryStageId = delivery.getCellValue(globalConfig.get('trigger_column'))?.id;
+
+    // assemble a list of existing message types for this delivery
+    let existingMessageTypes = [];
+    const existingMessages = delivery.getCellValue("Messages");
+    if (existingMessages) {
+      existingMessageTypes = existingMessages.map(existingMessage =>
+        findById(allMessages, existingMessage.id).getCellValue("Email type").name
+      );
+    } else {
+      existingMessageTypes = [];
+    }
+
+    // get the assigned volunteer, if any
+    let volunteer;
+    const volunteers = delivery.getCellValue("Assigned volunteer");
+    if (volunteers && volunteers.length > 0) {
+      volunteer = findById(allVolunteers, volunteers[0].id);
+    }
+
+    for (let messageType of Object.keys(EMAIL_TYPES)) {
+      if (existingMessageTypes.includes(messageType)) {
+        // We've already sent a message of this type for this delivery
+        continue;
+      }
+      if (EMAIL_TYPES[messageType].stage != deliveryStageId){
+        // The delivery is in the wrong stage for this type of email, skip
+        continue;
+      }
+
+      const {message, warning} = await formatMessageRecord(messageType, delivery, volunteer, EMAIL_TYPES[messageType]["audience"]);
+      if (message) {
+        output.messagesToCreate.push(message);
+      }
+      if (warning) {
+        output.warnings.push(warning);
+      }
+    }
+  }
+  return output;
+}
+
+
+function WarningAccordion({warnings}) {
+
+  if (warnings.length === 0) {
+    return null;
+  }
+  return (
+    <Accordion title={ `${warnings.length} warnings`}>
+      {warnings.map((warning, i) => {
+        return <p key={`warning_${i}`}>{warning}</p>
+      })}
+    </Accordion>
+  )
+}
+
+export function RefreshQueueStep({nextStep}) {
+  const [result, setResult] = useState(undefined);
+  const [warnings, setWarnings] = useState([]);
+
+  function addWarning(newWarning) {
+    if (newWarning) {
+      setWarnings(warnings.concat(newWarning));
+    }
+  }
+
+  async function refreshQueue() {
+    // Clear out the warnings from any previous run
+    setWarnings([]);
+    let {messagesToCreate, warnings} = await computeMessagesToCreate();
+    setWarnings(warnings);
+
+    if (messagesToCreate.length == 0) {
+      setResult(`No new messages to enqueue.`);
+      return;
+    }
+    if (messagesToCreate.length > 50) {
+      addWarning(`You can only create 50 emails at once, but we found ${messagesToCreate.length} Doing the first 50 now... remember to do this again to get the rest!`);
```

Comment:
I wonder if this error message could be more clear?

Reply:
Removed the need for this error message -- this is something we can just prevent from ever happening by batching, which is a better user experience than having to see this error and understand what to do